lexicon/providers/ddns.py

```diff
@@ -0,0 +1,145 @@
+
+    def _authenticate(self):
+        if not self.endpoint:
+            raise Exception("No DDNS server provided, use --ddns-server")
+        pass
+
+    # Create record. If record already exists with the same content, do nothing
+    def _create_record(self, rtype, name, content):
+        if self._get_lexicon_option("ttl"):
+            ttl = self._get_lexicon_option("ttl")
+        else:
+            ttl = 300
+        name = dns.name.from_text(name).relativize(dns.name.from_text(self.zone))
+
+        update = dns.update.Update(self.zone, keyring=self.keyring)
+        update.add(name, ttl, rtype, content)
+        self._run_query(update)
+        return True
+
+    # List all records. Return an empty list if no records found
+    # type, name and content are used to filter records.
+    # If possible filter during the query, otherwise filter after response is received.
+    def _list_records(self, rtype=None, name=None, content=None):
+        if rtype and name:  # doing a simple query is enough if we have type and name
+            query = dns.message.make_query(name, rtype)
+        else:  # if not, perform a zone transfert to get all records
+            query = dns.xfr.make_query(dns.versioned.Zone(self.zone), keyring=self.keyring)[0]
+
+        answers = self._run_query(query).answer
+
+        records = []
+        for answer in answers:  # filter unwanted results
+            a_name = answer.name.to_text().rstrip(".")
+            if name and name != a_name:
+                continue
+            a_rtype = dns.rdatatype.to_text(answer.rdtype)
+            if rtype and rtype != a_rtype:
+                continue
+            for rdata in answer:
+                rdata = rdata.to_text()
+                if content and content != rdata:
+                    continue
+                record = {
+                    "type": a_rtype,
+                    "name": a_name,
+                    "ttl": answer.ttl,
+                    "content": rdata,
+                    "id": f"{rtype}:{a_name}:{rdata}"
+                }
+                records.append(record)
+        LOGGER.debug("list_records: %s", records)
+        LOGGER.debug("Number of records retrieved: %d", len(records))
+        return records
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        if self._get_lexicon_option("ttl"):
+            ttl = self._get_lexicon_option("ttl")
+        else:
+            ttl = 300
+        if not identifier:
+            rrset = self._list_records(rtype, name)
+            if len(rrset) == 1:
+                identifier = rrset[0]["id"]
+            elif len(rrset) < 1:
+                raise Exception("No records found matching type and name - won't update")
+            else:
+                raise Exception("Multiple records found matching type and name - won't update")
+        d_rtype, d_name, d_content = identifier.split(":", 2)
```

Comment:
In general for the providers where identifiers are not provided by the API, I am more in favor of an identifier generated based on a hash, typically using this kind of function: https://github.com/AnalogJ/lexicon/blob/8caaeeffc14a0fb8fb4f32fba3935213f12c5187/lexicon/providers/azure.py#L361-L367

The reasons is to get a more usable id (something short, composed only of alphanumerical characters) and more importantly, an id that cannot ship any sensible data.

However this comes at a price: you cannot extract the actual record from the id. It means that in `_update_record` and `_delete_record`, when the `identifier` is used, you will need to list all records, recalculate their identifiers and select the record matching the `identifier` provided.

Is it fine for you?